### DISCUSSION TITLE
Add own LSP in a concise way?

### DISCUSSION THREAD

**Discussion by jcszymansk** [nix-community/nixvim#2011] [General]
Hi,

is it possible to add own LSP which is not supported by nixvim (and isn't even in `nixpkgs`) to the generated configuration? I can write a module from scratch, but is it possible to do something like this? :

```nix
  imports = [
    (mkLsp {
      name = "my-own-lsp";
      package = custompkgs.my-own-lsp-package;
      description = "My Own LSP";
      serverName = "ownlsp";
      cmd = cfg: [ (lib.getExe cfg.package) ];
    })
  ];

  programs.nixvim.plugins.lsp.servers.my-own-lsp.enable = true;
```

without copying and adapting `_mk-lsp.nix` or ugly hacks?

**Comment by MattSturgeon** [comment]
I think if we exposed our `modulesPath` you'd be able to `import "${modulesPath}/../plugins/lsp/language-servers/_mk-lsp.nix" { inherit pkgs lib etc; }`. Thinking about it, this is probably a best practice anyway...

But I don't _think_ you can easily do this currently.

That said, for a one-off language server, do you really need a `plugins.lsp.servers.*` module when you could probably just manually copy the implementation: https://github.com/nix-community/nixvim/blob/f13bdef0bc697261c51eab686c28c7e2e7b7db3c/plugins/lsp/language-servers/_mk-lsp.nix#L114-L133

**Comment by janekx21** [comment]
I also kind of want this for LSP development.

**Comment by jcszymansk** [comment]
This seems to work:

```nix
  mkLsp = lsp: {
    options.programs.nixvim = lib.mkOption {
      type = lib.types.submoduleWith {
        shorthandOnlyDefinesConfig = true;
        modules = [
          ((import "${inputs.nixvim}/plugins/lsp/language-servers/_mk-lsp.nix" { inherit lib config pkgs; }) lsp)
        ];
      };
    };
  };
```

**Reply by MattSturgeon** [reply]
It should be possible to simplify that to something like:

```nix
mkLspModule =
  args:
  # Return a module function
  { pkgs, lib, config, ... }:
  let
    mkLsp' = import "${inputs.nixvim}/plugins/lsp/language-servers/_mk-lsp.nix" { inherit pkgs lib config; };
  in
  { programs.nixvim = mkLsp' args; }
```

It shouldn't be necessary to extend the actual submodule option declaration.
